src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -106,17 +150,47 @@ def _transform_by_pattern(
                 tuple_constructor=lambda *elements: tuple(elements),
             )
 
+            def get_domain(
+                _, path: tuple[int, ...]
+            ) -> domain_utils.SymbolicDomain:  # function only used inside loop
+                domain = functools.reduce(
+                    lambda var, idx: var[idx] if isinstance(var, tuple) else var,
+                    path,
+                    tmp_expr.annex.domain,
+                )
+                assert isinstance(domain, domain_utils.SymbolicDomain)
+                return domain
+
+            tmp_domains: (
+                domain_utils.SymbolicDomain | tuple[domain_utils.SymbolicDomain | tuple, ...]
+            ) = type_info.apply_to_primitive_constituents(
+                get_domain,
+                tmp_expr.type,
+                with_path_arg=True,
+                tuple_constructor=lambda *elements: tuple(elements),
+            )
+
             # allocate temporary for all tuple elements
-            def allocate_temporary(tmp_name: str, dtype: ts.ScalarType):
-                declarations.append(itir.Temporary(id=tmp_name, domain=domain_expr, dtype=dtype))  # noqa: B023 # function only used inside loop
+            def allocate_temporary(
+                tmp_name: str, domain: domain_utils.SymbolicDomain, dtype: ts.ScalarType
+            ):
+                declarations.append(
+                    itir.Temporary(id=tmp_name, domain=domain.as_expr(), dtype=dtype)
+                )  # function only used inside loop
 
-            next_utils.tree_map(allocate_temporary)(tmp_names, tmp_dtypes)
+            next_utils.tree_map(allocate_temporary)(tmp_names, tmp_domains, tmp_dtypes)
```

Comment:
Not important, but this can be simplified by deleting the  `allocate_temporary()` function and instead using the constructor directly in `tree_map` which I also find slightly more readabale.
```suggestion
        declarations.extend(
            next_utils.tree_map(itir.Temporary)(
                tmp_names, tuple(d.as_expr() for d in tmp_domains), tmp_dtypes
            )
        )
```

Reply:
there is a flatten missing and the eve.Node only accepts kwargs, I propose
```
            declarations.extend(
                itir.Temporary(id=tmp_name, domain=domain.as_expr(), dtype=dtype)
                for tmp_name, domain, dtype in zip(
                    next_utils.flatten_nested_tuple(tmp_names),
                    next_utils.flatten_nested_tuple(tmp_domains),
                    next_utils.flatten_nested_tuple(tmp_dtypes),
                )
            )
```